### DISCUSSION TITLE
Rules for development

### DISCUSSION THREAD

**Discussion by BECATRUE** [snu-quiqcl/qiwis#2] [Announcements]
These are rules for development, especially about coding rules.

## Write docstring
- For easy development and management, write docstring with reference to [this](https://peps.python.org/pep-0257/).

## Snake case vs. Camel case
- For native Python codes, use `snake_case` as usual.
- For Qt-based codes, use `camelCase` for consistency with Qt modules.

## Type annotations
- Write type hints for all functions and methods arguments and its return type.
- See also: #78, #89

## Blank lines
- Follow [PEP8](https://peps.python.org/pep-0008/#blank-lines)
- Put a blank line right after a class docstring

## Imports
- Follow [PEP8](https://peps.python.org/pep-0008/#imports)
- In each group (standard, third-party, local), write `import` lines before `from ... import` lines

## Test for passing through Pyint
- This repository uses Pylint GitHub action for standardization of python code.
- You can test using Pylint like the below.
```
pylint ${file_path}.py         // test specific python file
pylint $(git ls-files '*.py')  // test all python files in git repositoy
```
- For more information about Pylint, look at [https://pylint.readthedocs.io/en/latest/](https://pylint.readthedocs.io/en/latest/).

### Initializing widgets
When you initialize a `QWidget`, do not separate the layout initialization part from `__init__()`.
In other words, do it in `__init__()`.
See #35 for more details.
